File: storage.py
import os
import sys
import socket
import subprocess


def is_nfs_server():
    ''' is localhost nfs server'''
    return os.path.isfile('/etc/exports')


def get_shares():
    # TODO: create resource class to associate with mounts and vm's
    ''' get local shares'''
    with open('/etc/exports', 'r') as f:
        exportsfile = f.readlines()
    exports = []
    for line in exportsfile:
        share = line_to_share(line)
        if share:
            exports.append(share)
    return exports


def get_mounts():
    ''' get nfs mounts on localhost '''
    mounts = []
    f = os.popen('df -h -t nfs -t nfs4 2> /dev/null')
    lines = f.read()
    lines = str(lines).splitlines()[1:]
    f.close()
    servername = socket.gethostname()
    for line in lines:
        line = line.split()
        server = line[0].split(':')
        path = server[1]
        server = server[0]
        name = path.split('/')[-1:][0]
        mount = {
            'mount': line[0],
            'shareserver': server,
            'sharepath': path,
            'sharename': name,
            'server': servername,
            'size': line[1],
            'used': line[2],
            'available': line[3],
            'usedPerc': line[4],
            'mountpoint': line[5]
        }
        mounts.append(mount)
    return mounts


def line_to_share(line):
    ''' line in exportfs to share dict'''
    if not line.strip()[0] == '#':
        servername = socket.gethostname()
        line = line.split()
        path = line[0]
        name = path.split('/')[-1:][0]
        options = line[1].split('(')
        network = options[0]
        options = options[1]
        options = options[:-1]
        options = options.split(',')
        share = {
            'name': name,
            'path': path,
            'network': network,
            'options': options,
            'server': servername
        }
        return share
    else:
        return False


def share_to_line(share):
    ''' share dict to exports line '''
    line = share['path'] + ' '
    line += share['network'] + '('
    for o in share['options']:
        line += o + ','
    line = line[:-1]
    line += ')\n'
    return line


def check_if_mounted(mountpoint):
    ''' check if mountpoint is mounted '''
    mounts = get_mounts()
    servername = socket.gethostname()
    for m in mounts:
        if m['mountpoint'] == mountpoint and m['server'] == servername:
            return True
    return False


def mount(share):
    ''' mount share '''
    mountpoint = '/mnt/'
    mountpoint += share['server'] + '/'
    mountpoint += share['name']
    share['mountpoint'] = mountpoint
    # print(mountpoint)
    if not check_if_mounted(mountpoint):
        # print('not mounted')
        sharepath = share['server'] + ':' + share['path']
        # print(sharepath)
        if not os.path.isdir(mountpoint):
            # print('create path' + str(mountpoint))
            cmd = ['mkdir', '-p', mountpoint]
            res = subprocess.call(cmd)
            if not res == 0:
                print('could not create mountpoint')
                print(str(sys.exc_info()[1]))
                return False
        cmd = ['mount', sharepath, mountpoint]
        # print('mounting')
        res = subprocess.call(cmd)
        if not res == 0:
            print('could not mount share')
            print(str(sys.exc_info()[1]))
            return False
        else:
            return share
    else:
        return share


def create_share(path):
    ''' create share from path on localhost'''
    # check if localhost is nfs server
    if not is_nfs_server():
        print('not a nfs server')
        return False
    # check if path exists
    if not os.path.isdir(path):
        print('path ' + path + ' does not exist')
        return False
    # check if already sahared
    shares = get_shares()
    for s in shares:
        if s['path'] == path:
            print('path already shared: ' + path)
            return False
    # add line to /etc/exports
    line = share_to_line({
        'path': path,
        'network': '10.0.0.1/24',
        'options': ['rw', 'sync', 'subtree_check', 'no_root_squash']
    })
    with open('/etc/exports', 'a') as f:
        f.write(line)
    # update exports
    res = subprocess.call(['exportfs', '-rav'])
    if res == 0:
        return True
    else:
        print('could not update exports')
        return False


def sync_share(primary_share, secondary_share):
    ''' sync primary share to secondary shares '''
    # mount shares
    primary_share = mount(primary_share)
    secondary_share = mount(secondary_share)
    if not primary_share:
        print('could not mount primary share')
        return False
    if not secondary_share:
        print('could not mount secondary share')
        return False
    prim_path = primary_share['mountpoint'] + '/'
    sec_path = secondary_share['mountpoint']
    cmd = ['rsync', '--archive', '--update', '--delete',
           '--exclude=.cloud.json', prim_path, sec_path]
    res = subprocess.call(cmd)
    if res == 0:
        return True
    else:
        return False


def migrate_share(share_name, to_server):
    ''' migrate share to server (sync, set primary) '''
    # TODO: resources; .cloud.json file in share to store metadata
    # TODO: migrate share
    pass

# arch: enable * start rpcbind.service + nfs-server.service
# exportfs -rav
# /nfs/test 10.0.0.1/24(rw,sync,subtree_check,no_root_squash)

File: storage/storage.go
package storage

import (
	// "fmt"
	"errors"

	"github.com/jmu0/cloud/functions"
	// "log"
	// "io"
	"os"
	"strings"
)

//storage struct for use with rpc
type Storage struct{}

//get shares, use with rpc
func (s *Storage) GetShares(opt string, shares *[]Share) error {
	var err error
	*shares, err = GetShares()
	return err
}

//get mounts, use with rpc
func (s *Storage) GetMounts(opt string, mounts *[]Mount) error {
	var err error
	*mounts, err = GetMounts()
	return err
}

//mount share on server
func (s *Storage) MountShare(sharestring string, reply *string) error {
	sh := Share{}
	err := sh.FromShareString(sharestring)
	if err != nil {
		return err
	}
	*reply = sharestring + " mounted"
	return sh.Mount()
}

//receive zfs dataset, use with rpc
func (s *Storage) ReceiveZfsSnapshot(stream DatasetStream, result *string) error {
	*result = "DEBUG: from ReceiveZfsSnapshot"
	return Receive(stream)
}

//structure to store properties for nfs share
type Share struct {
	Name    string
	Path    string
	Host    string
	Network string
	Options string
}

//get share fields from line in /etc/exports
func (s *Share) FromLine(line string) {
	tmp1 := strings.Fields(line)
	s.Path = tmp1[0]
	tmp2 := strings.Split(tmp1[1], "(")
	s.Network = tmp2[0]
	s.Options = tmp2[1][:len(tmp2[1])-1]
	s.Host, _ = functions.GetLocalhostName()
	if s.Path[len(s.Path)-1:] == "/" {
		s.Path = s.Path[:len(s.Path)-1]
	}
	tmp3 := strings.Split(s.Path, "/")
	s.Name = tmp3[len(tmp3)-1]
}

//get share fields from share string (server:path)
func (s *Share) FromShareString(share string) error {
	tmp1 := strings.Split(share, ":")
	if len(tmp1) != 2 {
		return errors.New("invalid share: " + share)
	}
	s.Host = tmp1[0]
	s.Path = tmp1[1]
	if s.Path[len(s.Path)-1:] == "/" {
		s.Path = s.Path[:len(s.Path)-1]
	}
	tmp2 := strings.Split(s.Path, "/")
	s.Name = tmp2[len(tmp2)-1]
	return nil
}

//create line for in /etc/exports
func (s *Share) ToLine() string {
	var line string
	line = s.Path
	if s.Network == "" {
		s.Network = "10.0.0.1/24"
	}
	if s.Options == "" {
		s.Options = "rw,sync,subtree_check,no_root_squash"
	}
	line += " " + s.Network
	line += "(" + s.Options + ")"
	return line
}

//mount share
func (s *Share) Mount() error {
	mountpoint := "/mnt/" + s.Host + "/" + s.Name
	share := s.Host + ":" + s.Path
	if functions.PathExists(mountpoint) == false {
		err := os.MkdirAll(mountpoint, 0770)
		if err != nil {
			return err
		}

	}
	// _, err := functions.ExecShell("mount", []string{share, mountpoint})
	_, err := functions.ExecShell("mount", share, mountpoint)
	if err != nil {
		return err
	}
	return nil
}

//structure to store properties for mounts
type Mount struct {
	Name       string
	Share      string
	SharePath  string
	ShareHost  string
	MountPoint string
	MountHost  string
}

//create mount from line in df
func (m *Mount) FromLine(line string) {
	fields := strings.Fields(line)
	m.Share = fields[0]
	tmp := strings.Split(m.Share, ":")
	m.ShareHost = tmp[0]
	m.SharePath = tmp[1]
	m.MountPoint = fields[5]
	m.MountHost, _ = functions.GetLocalhostName()
	tmp = strings.Split(m.Share, "/")
	m.Name = tmp[len(tmp)-1]
}

//returns if localhost is nfs server
func IsNfsServer() (bool, error) {
	// res, err := functions.ExecShell("pgrep", []string{"nfsd"})
	res, err := functions.ExecShell("pgrep", "nfsd")
	if err != nil {
		return false, nil
	} else {
		if len(res) > 0 {
			return true, nil
		}
	}
	return false, nil
}

//get shares on local server
func GetShares() ([]Share, error) {
	txt, err := functions.ReadFile("/etc/exports")
	if err != nil {
		return []Share{}, err
	}
	lines := strings.Split(txt, "\n")
	shares := []Share{}
	for _, l := range lines {
		if len(l) > 0 && l[:1] != "#" {
			s := Share{}
			s.FromLine(l)
			shares = append(shares, s)
		}
	}
	return shares, nil
}

//check if a path is shared
func IsShared(path string) (bool, error) {
	sh, err := GetShares()
	if err != nil {
		return false, err
	}
	for _, s := range sh {
		if s.Path == path {
			return true, nil
		}
	}
	return false, nil
}

//get mounts on local server
func GetMounts() ([]Mount, error) {
	mounts := []Mount{}
	// txt, err := functions.ExecShell("df", []string{"-h", "-t", "nfs", "-t", "nfs4"})
	txt, err := functions.ExecShell("df", "-h", "-t", "nfs", "-t", "nfs4")
	if err == nil {
		lines := strings.Split(txt, "\n")
		for _, l := range lines[1:] {
			if len(l) > 0 {
				m := Mount{}
				m.FromLine(l)
				mounts = append(mounts, m)
			}
		}
	}
	return mounts, nil
}

//create nfs share from folder
func CreateShare(path string) (Share, error) {
	if path[len(path)-1:] == "/" {
		path = path[:len(path)-1]
	}
	tmp := strings.Split(path, "/")
	host, _ := functions.GetLocalhostName()
	s := Share{
		Host: host,
		Name: tmp[len(tmp)-1],
		Path: path,
	}
	if ok, _ := IsNfsServer(); ok {
		if _, err := os.Stat(path); os.IsNotExist(err) {
			return s, errors.New("Path " + path + " does not exist")
		}
		isshared, err := IsShared(path)
		if err != nil {
			return s, err
		}
		if isshared {
			return s, errors.New(path + " is already shared")
		}
		err = functions.AppendToFile("/etc/exports", s.ToLine()+"\n")
		if err != nil {
			return s, err
		}
		// _, err = functions.ExecShell("exportfs", []string{"-rav"})
		_, err = functions.ExecShell("exportfs", "-rav")
		if err != nil {
			return s, err
		}

	} else {
		return s, errors.New("not a nfs server")
	}
	return s, nil
}
